### ISSUE TITLE
损失函数定义问题

### ISSUE THREAD

**Issue by OceanWong1991** [AITTSMD/MTCNN-Tensorflow#129]
@AITTSMD 您好！看original paper上损失函数 det, box, landmark loss 分别乘 ratio的和函数，看您的实现才用的是tf.reduce_mean()并没有才用tf.reduce_sum()，当测试时分别将计算损失函数的地方返回值改成reduce_sum(), 结果loss 均变成了nan,不知如何原因？能否指点一二，不胜感激！

**Comment by jetta** [comment]
请看`train.py`文件里调用`train_model`的地方。

**Comment by OceanWong1991** [comment]
@jetta 说明一下
**cls_loss_op,bbox_loss_op,landmark_loss_op**,L2_loss_op,accuracy_op = net_factory(input_image, label, bbox_target,landmark_target,training=True)
    #train,update learning rate(3 loss)
    train_op, lr_op = train_model(base_lr, radio_cls_loss*cls_loss_op + radio_bbox_loss*bbox_loss_op + radio_landmark_loss*landmark_loss_op + L2_loss_op, num)
   这三个变量在mtcnn_model.py landmark_ohem() bbox_ohem() ,cls_ohem() return  tf.reduce_mean()

**Comment by Dugudaibo** [comment]
@OceanWong1991 你好，我现在正在看 mtcnn_model.py 这部分，但是这个函数具体是如何计算的没有看懂，特别是计算cls_ohem()那部分，数组的shape变形和相加，已经被搞蒙了，请问可以讲一下大概的逻辑吗，就是先计算什么在计算什么之类的~

**Comment by OceanWong1991** [comment]
![cls_ohem](https://user-images.githubusercontent.com/17758711/40783601-8d25c1b8-6515-11e8-98c3-94d85c92f0ae.png)
@Dugudaibo 数据维度标注，应该不难理解了

**Comment by jjrCN** [comment]
@Dugudaibo 估计你是计算log部分没看懂，作者是把N*2的数组拉成2N长的向量，那么就相当于1个样本的两个softmax结果是纵向连续排列。比较巧妙的是通过indices_ = row + label_int这句话算出了标签为1的正样本在2N长度向量的位置，之后根据这个位置对本个batch中的所有样本取log，再之后，通过valid_inds找到当前batch中计算本次loss需要的样本，比如cls_ohem是需要0和-1标签样本，再通过loss = loss * valid_inds将不需要的样本log值变为0，最后loss,_ = tf.nn.top_k(loss, k=keep_num)
    return tf.reduce_mean(loss)取前70%loss大的并返回平均值。

**Comment by Dugudaibo** [comment]
我理解 indices_ = row + label_int 这句话不仅仅算出了标签为1 的位置，还算出了标签为 0 的位置，不知道对不对~ 
之后根据这个位置对本个batch中的所有样本取log：如果上面我理解的是对的话，那么这句话就是在计算 -y*log（p） - （1-y）*log（1-p），但是论文中给的公式是-y*log（p） - （1-y）*（1-log（p）），这两者的计算过程应该是不同的吧~ @jjrCN @AITTSMD

**Comment by jjrCN** [comment]
你说的对，确实不同，但是交叉熵损失函数是前面的，论文的应该是笔误。我是这样认为的。

**Comment by Dugudaibo** [comment]
我也是这么觉得的，笔误
好的好的，十分感谢~~~

也感谢上面的各位，还有开源的作者~~~
